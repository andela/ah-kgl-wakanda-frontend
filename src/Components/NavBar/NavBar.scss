@import '../../assets/sass/components';
@import '../../assets/sass/global';
.navbar {
  background: #fff;
  padding: 0.4em 0.8em !important;
  .search-container {
    margin-left: auto !important;
    width: 40%;
  }

  #dropdown-basic {
    .dropdown-menu {
      position: absolute;
      right: 3px !important;
      top: 57px !important;
    }
  }
}

@media #{$extra-small-device} {
  .navbar {
    padding: 0.4em 0 !important;
    .logo {
      display: none;
    }

    .search-container {
      width: 100%;
    }

    .searchbox {
      width: 90%;
      margin-left: 8px;
      padding-top: 4px;
    }

    .btn-auth {
      display: none;
    }

    .toggler-icon {
      color: $color-primary;
      border-radius: 4px;
      width: 38px;
      padding: 5px;
      height: 38px;
      cursor: pointer;

      &:hover {
        background: $color-secondary;
        color: $color-light;
      }

      &:active {
        background: $color-secondary;
        color: $color-light;
      }
    }
  }
}

@media #{$small-device} {
  .navbar {
    padding: 0.4em 0 !important;
    .logo {
      display: none;
    }

    .search-container {
      width: 100%;
    }

    .searchbox {
      width: 90%;
      margin-left: 8px;
      padding-top: 4px;
    }

    .btn-auth {
      display: none;
    }

    .toggler-icon {
      color: $color-primary;
      border-radius: 4px;
      width: 40px;
      padding: 5px;
      height: 42px;
      cursor: pointer;

      &:hover {
        background: $color-secondary;
        color: $color-light;
      }

      &:active {
        background: $color-secondary;
        color: $color-light;
      }
    }
  }
}

@media #{$medium-device} {
  .navbar {
    padding: 0.4em 0 !important;
    .logo {
      display: flex;
    }

    .search-container {
      margin-left: auto !important;
      width: 35%;
    }

    .searchbox {
      width: 90%;
      margin-left: 8px;
      padding-top: 4px;
    }

    .btn-auth {
      display: flex;
      flex-direction: row;
      margin-right: 2%;
    }

    .toggler-icon {
      display: none;
    }
  }
}

@media #{$large-device} {
  .navbar {
    .btn-auth {
      display: flex;
      flex-direction: row;
      margin-right: 2%;
    }

    .toggler-icon {
      display: none;
    }

    .logo {
      display: flex;
    }
  }
}

@media #{$extra-large-device} {
  .navbar {
    .btn-auth {
      display: flex;
      flex-direction: row;
      margin-right: 2%;
    }

    .toggler-icon {
      display: none;
    }
  }
}
