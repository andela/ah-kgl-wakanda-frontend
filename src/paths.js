export const HOME_PATH = '/';
export const CREATE_ARTICLE_PATH = '/articles/new';
export const SIGNIN_PATH = '/login';
export const SIGNUP_PATH = '/signup';
export const RESET_PASSWORD_PATH = '/reset-password';
export const UPDATE_PASSWORD_PATH = '/update-password/:token';
export const USER_ARTICLES = '/myarticles';
export const LOGOUT_PATH = '/logout';
export const ARTICLES_PATH = '/articles';
