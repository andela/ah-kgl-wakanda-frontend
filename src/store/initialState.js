const initialState = {
  name: { team: 'Wakanda' },
  title: 'Hello world',
};

export default initialState;
