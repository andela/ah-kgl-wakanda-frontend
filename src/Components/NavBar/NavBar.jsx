import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Navbar, Nav } from 'react-bootstrap';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBars } from '@fortawesome/free-solid-svg-icons';
import 'bootstrap/dist/css/bootstrap.css';
import './NavBar.scss';
import { toggleSideNav } from '../../actions/toggleSideNav';
import Logo from '../Common/Logo/Logo';
import NotifIcon from '../Common/NotifIcon/NotifIcon';
import SearchBox from '../Common/SearchBox/SearchBox';
import Button from '../Common/Button/Button';
import * as paths from '../../paths';
import ProfileDropdown from '../Common/Profile/ProfileDropdown';

/**
 * Render the NavBar
 * @class NavBar
 * @extends {Component}
 */
export class NavBar extends Component {
  /**
   * Creates an instance of NavBar
   * @param {*} props
   * @memberof NavBar
   */
  constructor(props) {
    super(props);
    const {
      navbar,
      onToggleSideNav,
      profile,
      currentUser,
      location,
      history,
      displaySearchBox,
    } = props;
    const { pathname } = location;
    this.profile = profile;
    this.displaySearchBox = displaySearchBox;
    this.pathname = pathname;
    this.history = history;
    this.navbar = navbar;
    this.onToggleSideNav = onToggleSideNav;
    this.currentUser = currentUser;
  }

  state = {
    searchText: '',
  };

  /**
   * help for typing in the searchBox
   * @param {Event} e
   * @memberof NavBar
   * @returns {void}
   */
  onChange = e => this.setState({ [e.target.name]: e.target.value });

  /**
   * @memberof NavBar
   * @returns {void}
   */
  toggleSideNav = () => {
    this.onToggleSideNav();
  };

  /**
   * TODO: the teammate who should implement notification feature
   * should complete this function
   * @param {Event} e
   * @returns {void}
   */
  toggleNotification(e) {
    e.preventDefault();
  }

  /**
   * render authentication buttons
   * @returns {object} jsx
   */
  renderAuthButton() {
    return (
      <Nav className="btn-auth ml-auto">
        <Nav.Link href={paths.SIGNUP_PATH}>
          <Button text="SIGN UP" />
        </Nav.Link>
        <Nav.Link href={paths.SIGNIN_PATH}>
          <Button text="SIGN IN" outline />
        </Nav.Link>
      </Nav>
    );
  }

  /**
   * render ProfileDropdown and Notification icon
   * @param {string} notificationsCount
   * @param {string} username
   * @param {string} image
   * @returns {object} jsx
   * @memberof NavBar
   */
  renderProfileNotifIcon(notificationsCount, username, image) {
    return (
      <Nav className="d-flex flex-row">
        {true || (
          <NotifIcon notificationsCount={notificationsCount} onClick={this.toggleNotification} />
        )}
        <ProfileDropdown username={username} picture={image} />
      </Nav>
    );
  }

  /**
   * @param {object} icon
   * @returns {object} jsx
   * @memberof NavBar
   */
  renderHamburgerIcon(icon) {
    return (
      <Nav className="ml-3">
        <FontAwesomeIcon
          onClick={this.toggleSideNav}
          className="toggler-icon"
          icon={icon}
          size="2x"
        />
      </Nav>
    );
  }

  /**
   * render navbar
   * @param {*} isAuth
   * @param {*} notificationsCount
   * @param {*} username
   * @param {*} image
   * @param {*} user
   * @returns {object} jsx
   */
  renderNavBar(isAuth, notificationsCount, username, image) {
    const { searchText } = this.state;
    return (
      <React.Fragment>
        <Navbar fixed="top" className="flex-nowrap flex-row" expand="lg" variant="dark">
          <Navbar.Brand className="d-none mr-md-auto d-md-block">
            <Logo />
          </Navbar.Brand>
          {this.renderHamburgerIcon(faBars)}
          {this.displaySearchBox ? (
            <Nav className="mr-auto search-container">
              <SearchBox history={this.history} value={searchText} onChange={this.onChange} />
            </Nav>
          ) : null}

          {!isAuth ? this.renderAuthButton() : null}
          {isAuth && this.pathname !== paths.CREATE_ARTICLE_PATH ? (
            <a href="/articles/new">
              <Nav className="ml-md-auto">
                <Button text="NEW ARTICLE" />
              </Nav>
            </a>
          ) : null}
          {!isAuth ? null : this.renderProfileNotifIcon(notificationsCount, username, image)}
        </Navbar>
      </React.Fragment>
    );
  }

  /**
   * render the component
   * @returns {object} jsx
   * @memberof Navbar
   */
  render() {
    const { isAuth, notification } = this.currentUser;
    const { notificationsCount } = notification;
    const {
      profile: { user },
    } = this.props;

    const { username, image } = user;
    return this.renderNavBar(isAuth, notificationsCount, username, image);
  }
}

/**
 * @param {object} state
 * @returns {object} props
 */
export const mapStateToProps = ({ navbar, profile, currentUser }) => ({
  navbar,
  profile,
  currentUser,
});

/**
 * @param {function} dispatch
 * @returns {object} props
 */
export const mapDispatchToProps = dispatch => ({
  onToggleSideNav: () => dispatch(toggleSideNav()),
});

NavBar.propTypes = {
  onToggleSideNav: PropTypes.func.isRequired,
  navbar: PropTypes.object.isRequired,
  currentUser: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  displaySearchBox: PropTypes.bool,
  profile: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
};

NavBar.defaultProps = {
  displaySearchBox: true,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(NavBar);
