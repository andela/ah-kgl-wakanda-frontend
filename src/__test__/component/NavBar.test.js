import React from 'react';
import { shallow } from 'enzyme';
import { createMemoryHistory } from 'history';
import { NavBar, mapDispatchToProps, mapStateToProps } from '../../Components/NavBar/NavBar';
import initialState from '../../store/initialState';
import { HOME_PATH } from '../../paths';

describe('NavBar component', () => {
  let props;
  let wrapper;
  const e = { preventDefault: jest.fn() };
  const history = createMemoryHistory();
  beforeEach(() => {
    props = {
      onToggleSideNav: jest.fn(),
      navbar: initialState.navbar,
      currentUser: {
        isAuth: true,
        user: {
          username: 'test',
        },
        notification: {
          notificationList: [],
          notificationsCount: 0,
        },
      },
      profile: initialState.profile,
      location: { pathname: HOME_PATH },
      history,
      displaySearchBox: true,
      onClick: jest.fn(e),
      onChange: jest.fn(e),
      onSearch: jest.fn(e),
    };
    wrapper = shallow(<NavBar {...props} />);
  });

  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should map dispatch to props', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).onToggleSideNav();
    expect(dispatch).toHaveBeenCalled();
  });

  it('should execute instance methods', () => {
    const isDrawerDisplay = false;
    const instance = wrapper.instance();
    instance.toggleSideNav(isDrawerDisplay);
    expect(props.onToggleSideNav).toHaveBeenCalled();
  });

  it('should map state to props', () => {
    const navbar = {
      isDrawerDisplay: false,
    };
    expect(mapStateToProps({ navbar })).toBeDefined();
  });

  it('should set state `value`', () => {
    const searchText = 'Title for testing';
    wrapper.setState({
      searchText,
    });

    wrapper.find('Connect(SearchBox)').simulate('change', { target: { value: searchText } });
    expect(wrapper.state('searchText')).toBe(searchText);
  });

  it('should display auth button', () => {
    const newProps = {
      ...props,
      displaySearchBox: false,
      currentUser: {
        isAuth: false,
        user: {
          username: 'test',
        },
        notification: {
          notificationList: [],
          notificationsCount: 0,
        },
      },
    };

    wrapper = shallow(<NavBar {...newProps} />);
  });
});
