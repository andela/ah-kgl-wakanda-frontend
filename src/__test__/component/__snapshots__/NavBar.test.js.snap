// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar component should render without crashing 1`] = `ShallowWrapper {}`;
